### PR TITLE
Create Zone Hold functionality

### LINKED ISSUE #1364: Cloudflare Zone Hold Support
### Current cloudflare-go version

v0.74.0

### Description

Add sdk support for creating/deleting holds:

https://developers.cloudflare.com/api/operations/zones-0-hold-post
https://developers.cloudflare.com/api/operations/zones-0-hold-get

### Use cases

Add or remove zone holds via sdk

### Potential cloudflare-go usage

Interact with the API via:
https://github.com/cloudflare/terraform-provider-cloudflare/pull/2671

### References

_No response_

### PR DESCRIPTION
**Pull request by danquack**
## Description

Adds the ability to create a zone hold based on the hold API.

## Has your change been tested?

I called the apis manually. I am happy to write tests, but I'm not sure where to get started. I could stub the responses, but since this is basically just calling the raw api, im not sure what I would be testing (no additional logic).

## Types of changes

What sort of change does your code introduce/modify?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] This change is using publicly documented in [cloudflare/api-schemas](https://github.com/cloudflare/api-schemas) 
      and relies on stable APIs.

[1]: https://help.github.com/articles/closing-issues-using-keywords/
closes #1364

### CODE REVIEW

File: hold.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/goccy/go-json"
+)
+
+// ZoneHold represents a Zone Hold rule. A rule only permits access to
```
Comment (patryk) [inline hold.go]: The comment doesn't match the expected behaviour. You may use copy found on https://developers.cloudflare.com/api/operations/zones-0-hold-post.

File: hold.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+// endpoint.
+type ZoneHoldDeleteResponse struct {
+	Result ZoneHold `json:"result"`
+}
+
+// ZoneHoldCreateParams contains required and optional params
+// for creating a zone Hold.
+type ZoneHoldCreateParams struct {
+	ID                string `json:"id"`
+	Hold              bool   `json:"hold"`
```
Comment (patryk) [inline hold.go]: These two fields are not defined for POST request.

File: hold.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	IncludeSubdomains *bool  `json:"include_subdomains,omitempty"`
+}
+
+// ZoneHoldDeleteParams contains required and optional params
+// for updating a zone Hold.
+type ZoneHoldDeleteParams struct {
+	ID        string  `json:"id"`
+	Hold      bool    `json:"hold"`
```
Comment (patryk) [inline hold.go]: Same.

File: hold.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+// ZoneHoldResponse represents a response from the Zone Hold endpoint.
+type ZoneHoldResponse struct {
+	Result ZoneHold `json:"result"`
+	Response
+	ResultInfo `json:"result_info"`
+}
+
+// ZoneHoldDeleteResponse represents a response from the List Zone Hold
```
Comment (tpaulus) [inline hold.go]: Comment describes a List Zone Hold endpoint, which does not exist.

File: hold.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+// endpoint.
+type ZoneHoldDeleteResponse struct {
+	Result ZoneHold `json:"result"`
+}
+
+// ZoneHoldCreateParams contains required and optional params
```
Comment (tpaulus) [inline hold.go]: There is only 1 parameter (`include_subdomains`) for Create Zone Hold Request, and it is optional.

File: hold.go (hunk 1 above)
Comment (danquack) [reply hold.go]: A lot of these were copy pasta from a different file. I updated not, can you check if that works for you.

File: hold.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+// for creating a zone Hold.
+type ZoneHoldCreateParams struct {
+	ID                string `json:"id"`
+	Hold              bool   `json:"hold"`
+	IncludeSubdomains *bool  `json:"include_subdomains,omitempty"`
+}
+
+// ZoneHoldDeleteParams contains required and optional params
```
Comment (tpaulus) [inline hold.go]: There is only 1 parameter (`hold_after`) for a Delete Zone Hold request and it is optional.

File: hold.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+// for updating a zone Hold.
+type ZoneHoldDeleteParams struct {
+	ID        string  `json:"id"`
+	Hold      bool    `json:"hold"`
+	HoldAfter *string `json:"hold_after,omitempty"`
```
Comment (tpaulus) [inline hold.go]: This must be an RFC3339-formatted timestamp, should the type be changed to `time` and this conversion done as part of the Delete Operation to ensure operation performs as expected?

File: hold.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+// CreateZoneHold creates a Zone ZoneHold rule for the given zone ID.
+//
+// API reference: https://developers.cloudflare.com/api/operations/zones-0-hold-post
+func (api *API) CreateZoneHold(ctx context.Context, zoneID string, params ZoneHoldCreateParams) (ZoneHold, error) {
+	uri := fmt.Sprintf("/zones/%s/hold", zoneID)
+	if params.IncludeSubdomains != nil {
+		uri = fmt.Sprintf("%s?include_subdomains=%t", uri, *params.IncludeSubdomains)
```
Comment (tpaulus) [inline hold.go]: Thoughts on using `url` package's `Values` type for this encoding?

https://stackoverflow.com/a/36000522/8156448

File: hold.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+	return response.Result, nil
+}
+
+// DeleteZoneHold deletes a Zone ZoneHold rule (based on the ID) for the given zone ID.
+//
+// API reference:https://developers.cloudflare.com/api/operations/zones-0-hold-delete
+func (api *API) DeleteZoneHold(ctx context.Context, zoneID string, params ZoneHoldDeleteParams) (ZoneHold, error) {
+	uri := fmt.Sprintf("/zones/%s/hold", zoneID)
+	if params.HoldAfter != nil {
+		uri = fmt.Sprintf("%s?hold_after=%s", uri, *params.HoldAfter)
```
Comment (tpaulus) [inline hold.go]: Thoughts on using `url` package's `Values` type for this encoding?

https://stackoverflow.com/a/36000522/8156448

**Review by tpaulus — CHANGES_REQUESTED** [review]

File: hold.go (hunk 7 above)
Comment (danquack) [reply hold.go]: converted.

File: hold.go (hunk 9 above)
Comment (danquack) [reply hold.go]: Id be fine either way, my thought to use sprintf was just because it was a total of one param

**Review by tpaulus — APPROVED** [review]

File: hold.go (hunk 8 above)
Comment (jacobbednarz) [reply hold.go]: we have [`buildURI`](https://github.com/cloudflare/cloudflare-go/blob/80b7949c7f0b8f051b7b5e79f6272f30c552c00c/utils.go#L13-L16) for this purpose. it allows adding query parameter tags to the struct without worrying about manually building path + query parameters.

example: https://github.com/cloudflare/cloudflare-go/blob/80b7949c7f0b8f051b7b5e79f6272f30c552c00c/tunnel_routes.go#L85

File: hold.go [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+// ZoneHoldDeleteParams represents params for the Delete Zone Hold
+// endpoint.
+type ZoneHoldDeleteParams struct {
+	HoldAfter *time.Time `json:"hold_after,omitempty"`
+}
+
+// Enforce a zone hold on the zone, blocking the creation and activation of zones with this zone's hostname.
+//
+// API reference: https://developers.cloudflare.com/api/operations/zones-0-hold-post
+func (api *API) CreateZoneHold(ctx context.Context, zoneID string, params ZoneHoldCreateParams) (ZoneHold, error) {
```
Comment (jacobbednarz) [inline hold.go]: we need to update the method signature here to match the library conventions - https://github.com/cloudflare/cloudflare-go/blob/master/docs/conventions.md#methods

File: hold.go [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+		return ZoneHold{}, fmt.Errorf("%s: %w", errUnmarshalError, err)
+	}
+
+	return response.Result, nil
+}
+
+// Retrieve whether the zone is subject to a zone hold, and metadata about the hold.
+//
+// API reference: https://developers.cloudflare.com/api/operations/zones-0-hold-get
+func (api *API) ZoneHold(ctx context.Context, zoneID string) (ZoneHold, error) {
```
Comment (jacobbednarz) [inline hold.go]: updating this to follow the library  conventions for CRUD naming

```suggestion
func (api *API) GetZoneHold(ctx context.Context, zoneID string) (ZoneHold, error) {
```

we'll also need to update the signature to match https://github.com/cloudflare/cloudflare-go/blob/master/docs/conventions.md#methods

**Review by jacobbednarz — CHANGES_REQUESTED** [review]
thanks for the PR! 

i've left some feedback inline about the library conventions we need to update here but we'll also need to add test coverage. you can check out another file like https://github.com/cloudflare/cloudflare-go/blob/master/accounts_test.go to get an idea on the minimums we expect here.

File: hold.go (hunk 10 above)
Comment (danquack) [reply hold.go]: Questions on the ResourceContainer. I copied my functions from here? Are they outside convention?

https://github.com/cloudflare/cloudflare-go/blob/64904badcb5202c3ad02d4a1c90bbfa7432f9636/firewall.go#L117

if I switch to the container, where does zone ID come from? Does it come from the container or do I have to add it as a param?

Comment (jacobbednarz) [reply hold.go]: yes, that is an older method that has not yet been converted to the patterns outlined in the conventions and [experimental client doc](https://github.com/cloudflare/cloudflare-go/blob/master/docs/experimental.md). you're best off checking out a more modern file like https://github.com/cloudflare/cloudflare-go/blob/master/rulesets.go#L758 to see how it works.

> if I switch to the container, where does zone ID come from? Does it come from the container or do I have to add it as a param?

`*ResourceContainer` is the basic way of describing a resource and its identifier (https://github.com/cloudflare/cloudflare-go/blob/master/resource.go). the method accepts this as a parameter but end usage should use the helper method like [`ZoneIdentifier`](https://github.com/cloudflare/cloudflare-go/blob/a0f7a6b7de3d9023ef32fa9f518b94e4d68e3317/resource.go#L99-L105) which handles creating the struct and the associated relationships for you.

**Pull request merged**